### PR TITLE
`exports.default` in ES5 vs `exports["default"]` in ES3

### PR DESCRIPTION
**Pull request by truongsinh**
I am not sure if I am supposed to run `jake LKG`, but I did it anyway, so I can immediately use this `tsc` in my project.

There are 4 new test cases to cover this scenario; some other test cases are just updated to have `target: ES5`, as I believed that was author's original idea.

### CODE REVIEW

**Comment by truongsinh** [comment]
CLA signed https://www.docusign.net/Signing/?ti=0bc2cb71ced04773b2b6796a65fbbc39

**Comment by mhegazy** [comment]
this looks related to #2387. pinging @billti.

**Comment by billti** [comment]
Yes, see the code change in the issue Mohamed linked to.  There are a bunch of keywords reserved in ES3 but not ES5/6.  We shouldn't special case just the one ('default').

This is also a deeper issue than just changing member access from dot to index notation.  For example, anywhere we emit an identifier (i.e. internal module name, class name, etc...), we need to prohibit these additional reserved words as names if emitting for ES3.

There is a closely related issue with there being different reserved words in strict mode (and with ES6 modules & classes being strict by default). I believe @yuit was looking into this.

I have a code review on the issue referenced, but I need to update it since I identified these additional concerns.  Sorry it's taking a while.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11412,7 +11412,14 @@ module ts {
             let node = getDeclarationOfAliasSymbol(symbol);
             if (node) {
                 if (node.kind === SyntaxKind.ImportClause) {
-                    return getGeneratedNameForNode(<ImportDeclaration>node.parent) + ".default";
+                    let defaultKeyword: string;
+
+                    if (compilerOptions.target == ScriptTarget.ES3) {
```
Comment (yuit) [inline src/compiler/checker.ts]: Use this check instead ```languageVersion === ScriptTarget.ES3```

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -2452,7 +2452,11 @@ var __param = this.__param || function(index, decorator) { return function (targ
                     writeLine();
                     emitStart(node);
                     if (node.flags & NodeFlags.Default) {
-                        write("exports.default");
+                        if (compilerOptions.target == ScriptTarget.ES3) {
```
Comment (yuit) [inline src/compiler/emitter.ts]: Same here; use this ```languageVersion === ScriptTarget.ES3```

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -4548,7 +4552,11 @@ var __param = this.__param || function(index, decorator) { return function (targ
                         writeLine();
                         emitStart(node);
                         emitContainingModuleName(node);
-                        write(".default = ");
+                        if (compilerOptions.target == ScriptTarget.ES3) {
```
Comment (yuit) [inline src/compiler/emitter.ts]: same here

**Comment by yuit** [comment]
I agree with @billti  that this will be more preferable if it is not a special case to "default". To extend this for ES3, you will have to add ES3 reserved word (that is not already ES5/6 reserved word- i.e synchronized, transient, volatile, byte etc) into the SyntaxKind. In the most recent compiler, when an identifier get created and the token is not really an SyntaxKind.Identifier, the property call ```originalKeywordKind``` will be set. [See This:] (https://github.com/Microsoft/TypeScript/blob/master/src/compiler/parser.ts#L1491). Then, you would want to check this property and emit using index notation instead of identifier for target ES3.

**Comment by DanielRosenwasser** [comment]
Please remove changes to the LKG; the team is responsible for updating it.

**Comment by truongsinh** [comment]
@yuit @billti is there any ETA for #2387? Because breaking ES3 (at least with modules) is a critical problem for my project. I will update `languageVersion === ScriptTarget.ES3` and remove LKG as suggested anyway.

**Comment by mhegazy** [comment]
@truongsinh can you update the PR and remove the bin\* files

**Comment by truongsinh** [comment]
Done, sorry for the procrastination.

**Comment by mhegazy** [comment]
merged. thanks @truongsinh

**Pull request merged**